src/Sparrow.Server/Platform/Posix/CGroupHelper.cs

```diff
@@ -0,0 +1,315 @@
+            Time = time;
+        }
+
+        public void Deprecate()
+        {
+            Time = DateTime.MinValue;
+        }
+    }
+
+    protected CGroup()
+    {
+        _groupPathForMemory = CreateNewLazyCachedPath();
+    }
+    
+    public long? GetMaxMemoryUsage() => ReadValue(GetGroupPathForMemory, MaxMemoryUsageFileName);
+    public long? GetPhysicalMemoryUsage() => ReadValue(GetGroupPathForMemory, MemoryUsageFileName);
+    public long? GetPhysicalMemoryLimit() => ReadValue(GetGroupPathForMemory, MemoryLimitFileName, CheckLimitValues);
+    
+    private Lazy<CachedPath> GetGroupPathForMemory()
+    {
+        var groupPathForMemory = _groupPathForMemory;
+        if (DateTime.UtcNow - groupPathForMemory.Value.Time > TimeSpan.FromMinutes(1))
+        {
+            var newVal = CreateNewLazyCachedPath();
+            Interlocked.CompareExchange(ref _groupPathForMemory, newVal, groupPathForMemory);
+        }
+
+        return _groupPathForMemory;
+    }
+
+    private Lazy<CachedPath> CreateNewLazyCachedPath()
+    {
+        return new Lazy<CachedPath>(() =>
+        {
+            string path = null;
+            try
+            {
+                path = FindCGroupPathForMemory();
+            }
+            catch (Exception e)
+            {
+                if (Logger.IsOperationsEnabled)
+                    Logger.Operations("Failed to get CGroup path for memory", e);
+            }
+
+            return new CachedPath(path, DateTime.UtcNow);
+        });
+    }
+    
+    private static bool CheckLimitValues(string textValue, out long? value)
+    {
+        //'max' stand for unlimited 
+        if (textValue.StartsWith("max"))
+        {
+            value = long.MaxValue;
+            return true;
+        }
+
+        value = null;
+        return false;
+    }
+
+    private static long? ReadValue(Func<Lazy<CachedPath>> getBasePath, string file, CheckSpecialValues checkSpecialValues = null, bool retry = true)
+    {
+        var basePath = getBasePath();
+        if (basePath.Value.Path == null)
+            return null;
+
+        try
+        {
+            var path = Path.Combine(basePath.Value.Path, file);
+            return ReadMemoryValueFromFile(path, checkSpecialValues);
+        }
+        catch (Exception e)
+        {
+            if (e is not DirectoryNotFoundException || retry == false) 
+                return null;
+            
+            //If the cgroup changed and the old cgroup was removed
+            basePath.Value.Deprecate();
+            return ReadValue(getBasePath, file, checkSpecialValues, false);
+        }
+    }
+
+    protected abstract string FindCGroupPathForMemory();
+    
+    //51 34 0:46 / /sys/fs/cgroup/hugetlb rw,nosuid,nodev,noexec,relatime shared:27 - cgroup cgroup rw,hugetlb
+    private static readonly Regex FindHierarchyMountReg = new Regex(@"^(?:\S+\s+){3}(?<mountroot>\S+)\s+(?<mountpath>\S+).* - (?<filesystemType>\S+)\s+\S+\s+(?:(?<options>[^,]+),?)+$");
```

Comment:
This should probably be compiled

Reply:
Sorry, I am not sure what you mean

Reply:
https://learn.microsoft.com/en-us/dotnet/standard/base-types/compilation-and-reuse-in-regular-expressions

Reply:
That is very nice.
I added the flag